rfcs/0100-sign-commits.md

```diff
@@ -0,0 +1,345 @@
+- https://git-scm.com/docs/shallow
+
+We likely want to make Nix use `--filter=tree:0` by default when fetching repositories.
+`--depth 1` is faster, but it doesn't preserve the commit history,
+and it's not clear how it it interacts with PGP signing.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+Committers will need to adapt to the new workflow before March of 2022.
+The command they will run is explained in [signing], but further
+information about how to sign commits is out of scope for this RFC.
+
+Relevant documentation for setting up GPG:
+- https://guix.gnu.org/manual/en/html_node/Commit-Access.html
+- https://www.kernel.org/doc/html/latest/process/maintainer-pgp-guide.html
+
+Flake users will need to become aware of the basics of this system
+when using forks of Nixpkgs, as updating such flake inputs
+will be different from before.
+In addition, flake users should also begin taking use of the features
+introduced in this RFC, as they are usable and useful for all flakes.
+
+## Example flake
+
+```nix
+# flake.nix
+{
+  inputs.someflake = {
+    url = "https://github.com/some/flake.git";
+    startsAt = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";
+    startsAtKey = "XXXX XXXX XXXX XXXX XXXX  XXXX XXXX XXXX XXXX XXXX";
+    type = "git";
+  };
+
+  outputs = { someflake }: {};
+}
+```
+
+```nix
+# some/flake/authorized-committers.nix
+{
+  committers.some = "XXXX XXXX XXXX XXXX XXXX  XXXX XXXX XXXX XXXX XXXX";
+}
+```
+
+Now if `https://github.com/some/flake.git` receives a new commit that can
+*not* be trusted according to the algorithm specified in [updating],
+we'd get an error like the following if we run `nix flake update`:
+```
+* **Failed trust verification!** 'someflake': 'git+https://github.com/some/flake.git?ref=master&rev=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' -!> 'git+https://github.com/some/flake.git?ref=master&rev=bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb' (Not signed by a trusted party!)
+```
+
+Notably, there is nothing that tells the user how to bypass the check
+using the `insecureUpdate` attribute,
+since if there were, users might very often ignore the error even if valid.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Merging PRs can not be done through GitHub.
+- Setting up PGP signatures will be a bit extra work for the committers.
+- Committers will need to maintain the list of authorized committers.
+- Fetching Nixpkgs with Nix will take around [5 times](#speeding-up-fetching-git) the time, depending on your hardware.
+- Currently building a fork of Nixpkgs (for e.g. a PR) will still work fine, since
+  the latest commit will be trusted when adding a new flake input, or referring
+  to one from the commandline, but updating the flake input will fail
+  since the commits are likely not signed by authorized committers.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Mandating signing only tags is an alternative, though not very useful for most users, since:
```

Comment:
We don't even set tags outside of the initial release process. There are no point-updates or similar in NixOS. Tags are basically useless for us.

Reply:
I realize this, but I put it in the alternatives part to explain why it *isn't* an alternative. I'll modify my wording a bit.